### DISCUSSION TITLE
Using global EntityManager instance methods for context specific actions is disallowed.

### DISCUSSION THREAD

**Discussion by n4an** [mikro-orm/mikro-orm#2531] [Q&A] [answered]
Error occur when in service I create an entity

Using global EntityManager instance methods for context specific actions is disallowed. If you need to work with the global instance's identity map, use `allowGlobalContext` configuration option or `fork()` instead.

**Comment by B4nan** [comment]
> to add private readonly orm: MikroORM to service or controller?

I am having hard times understanding your english... What about using full sentences? You know, subject, verb, object :]

That validation error is throw if you don't have request context set up correctly, or in general when you try to work with global entity manager, outside of request scope handler. In nest the request context is set up automatically, unless you disable it via `registerRequestContext: false` (the default is true).

Ideally create a reproduction project, will be happy to look into it. The realworld example project works fine in this sense, so you can try to compare what you have to that.

**Reply by B4nan** [reply]
Yes, that is my point, by default it is enabled. Only `false` value will disable it, not `undefined`.

**Comment by n4an** [comment]


**Reply by B4nan** [reply]
I am not able to run that project

```
➜  nest-mikro-rom git:(main) ✗ yarn nest:dev
yarn run v1.22.15
$ nest start --watch
[19:46:34] Starting compilation in watch mode...

src/modules/users/users.controller.ts:11:4 - error TS2304: Cannot find name 'Post'.

11   @Post('test')
      ~~~~

src/modules/users/users.controller.ts:13:6 - error TS2693: 'Body' only refers to a type, but is being used as a value here.

13     @Body() createUserDto: any
        ~~~~

src/modules/users/users.controller.ts:14:14 - error TS2304: Cannot find name 'User'.

14   ): Promise<User> {
                ~~~~

[19:46:35] Found 3 errors. Watching for file changes.
```

**Reply by B4nan** [reply]
It works fine on my end. I did clean things up a bit and used sqlite instead, you had few configuration options that are not available, like `synchronize` or `logging`, but I don't think that should have any effect. I am also testing via get request as I am too lazy to start tools like postman :]

I added logging of the EM id to the user service to ensure it really does use fresh forks, and indeed it does:

```
[20:09:49] Found 0 errors. Watching for file changes.

[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [NestFactory] Starting Nest application...
[discovery] ORM entity discovery started, using ReflectMetadataProvider
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [InstanceLoader] AppModule dependencies initialized +22ms
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [InstanceLoader] MikroOrmModule dependencies initialized +0ms
[discovery] - processing 1 files
[discovery] - processing entity User (/Users/adamek/htdocs/mikro-orm/nest-mikro-rom/dist/modules/users/entities/user.entity.js)
[discovery] - processing entity BaseEntity
[discovery] - entity discovery finished, found 2 entities, took 13 ms
[info] MikroORM successfully connected to database test.db
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [InstanceLoader] MikroOrmCoreModule dependencies initialized +43ms
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [InstanceLoader] MikroOrmModule dependencies initialized +1ms
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [InstanceLoader] UsersModule dependencies initialized +0ms
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [RoutesResolver] UserController {/api/user}: +3ms
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [RouterExplorer] Mapped {/api/user, POST} route +3ms
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [RoutesResolver] UsersController {/api/users}: +0ms
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [RouterExplorer] Mapped {/api/users/test, POST} route +1ms
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [RouterExplorer] Mapped {/api/users/test, GET} route +0ms
[Nest] 88766  - 14. 12. 2021 20:09:49     LOG [NestApplication] Nest application successfully started +2ms
{"level":30,"time":1639508989684,"pid":88766,"hostname":"Martin-MacBook-Pro.local","msg":"Server listening at http://127.0.0.1:3000"}
{"level":30,"time":1639508990972,"pid":88766,"hostname":"Martin-MacBook-Pro.local","reqId":"req-1","req":{"method":"GET","url":"/api/users/test","hostname":"127.0.0.1:3000","remoteAddress":"127.0.0.1","remotePort":63533},"msg":"incoming request"}
3
[query] begin
[query] insert into `user` (`email`, `first_name`, `last_name`, `password`) values ('e-0.29731642177696926', 'a', 'b', '$2b$08$4uC2YvqS9m7rWo08dSnTcOO9vO1/ESQBPKxbCbGYR73xx8glW0ItC') [took 2 ms]
[query] commit
{"level":30,"time":1639508991033,"pid":88766,"hostname":"Martin-MacBook-Pro.local","reqId":"req-1","res":{"statusCode":200},"responseTime":60.33104205131531,"msg":"request completed"}
{"level":30,"time":1639508991434,"pid":88766,"hostname":"Martin-MacBook-Pro.local","reqId":"req-2","req":{"method":"GET","url":"/api/users/test","hostname":"127.0.0.1:3000","remoteAddress":"127.0.0.1","remotePort":63533},"msg":"incoming request"}
4
[query] begin
[query] insert into `user` (`email`, `first_name`, `last_name`, `password`) values ('e-0.982862438936974', 'a', 'b', '$2b$08$PP8hnmDG3zAo4SC7vzgf8.asG0aumx8gBn1D2S6pqzGklE1.cZpry') [took 1 ms]
[query] commit
{"level":30,"time":1639508991471,"pid":88766,"hostname":"Martin-MacBook-Pro.local","reqId":"req-2","res":{"statusCode":200},"responseTime":37.073832988739014,"msg":"request completed"}
{"level":30,"time":1639508991805,"pid":88766,"hostname":"Martin-MacBook-Pro.local","reqId":"req-3","req":{"method":"GET","url":"/api/users/test","hostname":"127.0.0.1:3000","remoteAddress":"127.0.0.1","remotePort":63533},"msg":"incoming request"}
5
[query] begin
[query] insert into `user` (`email`, `first_name`, `last_name`, `password`) values ('e-0.8741729936499703', 'a', 'b', '$2b$08$IGWI1bvMW9LznAltu6IYz.YqWnJDW8ByxKlzg.dS0uKpz6uGBV5RS') [took 1 ms]
[query] commit
{"level":30,"time":1639508991834,"pid":88766,"hostname":"Martin-MacBook-Pro.local","reqId":"req-3","res":{"statusCode":200},"responseTime":28.182250022888184,"msg":"request completed"}
```

Here are the changes I have done on my local copy before making it work:

```diff
diff --git a/package.json b/package.json
index 2f295f8..f189a5b 100755
--- a/package.json
+++ b/package.json
@@ -23,7 +23,7 @@
   "dependencies": {
     "@mikro-orm/core": "^5.0.0-dev.562",
     "@mikro-orm/migrations": "^5.0.0-dev.562",
-    "@mikro-orm/mysql": "^5.0.0-dev.562",
+    "@mikro-orm/sqlite": "^5.0.0-dev.562",
     "@mikro-orm/nestjs": "^4.3.1",
     "@mikro-orm/reflection": "^5.0.0-dev.562",
     "@mikro-orm/sql-highlighter": "^1.0.1",
diff --git a/src/app.module.ts b/src/app.module.ts
index df148af..9c21a13 100644
--- a/src/app.module.ts
+++ b/src/app.module.ts
@@ -8,33 +8,21 @@ import { UsersModule } from './modules/users/users.module'
   imports: [
     MikroOrmModule.forRootAsync({
       useFactory: () => ({
-        autoLoadEntities: false,
-        dbName: 'nuxt_nest',
-        debug: ['discovery', 'info', 'query', 'query-params'],
+        // autoLoadEntities: false,
+        dbName: 'test.db',
+        debug: true,
         discovery: {
           warnWhenNoEntities: true
         },
-        host: 'localhost',
         logging: true,
-        password: '123456',
         // metadataProvider: TsMorphMetadataProvider,
         port: 3306,
-        type: 'mysql',
-        username: 'root',
-        synchronize: true,
+        type: 'sqlite',
         entities: [__dirname + '/modules/**/entities/*.entity.js'],
         entitiesTs: [__dirname + '/modules/**/entities/*.entity.ts'],
-        logger: msg => (new Logger('MikroOrm')).log(msg)
       })
     }),
     UsersModule
   ]
 })
-export class AppModule implements NestModule {
-  configure(consumer: MiddlewareConsumer) {
-    consumer
-      .apply(MikroOrmMiddleware)
-      .forRoutes('(.*)')
-      // .forRoutes('*')
-  }
-}
+export class AppModule { }
diff --git a/src/modules/users/repositories/user.repository.ts b/src/modules/users/repositories/user.repository.ts
index 129322b..b2e7e1a 100644
--- a/src/modules/users/repositories/user.repository.ts
+++ b/src/modules/users/repositories/user.repository.ts
@@ -1,4 +1,4 @@
-import { EntityRepository } from '@mikro-orm/mysql'
+import { EntityRepository } from '@mikro-orm/sqlite'
 
 import { User } from '../entities/user.entity'
 
diff --git a/src/modules/users/users.controller.ts b/src/modules/users/users.controller.ts
index 6e8e381..dc9e385 100644
--- a/src/modules/users/users.controller.ts
+++ b/src/modules/users/users.controller.ts
@@ -1,6 +1,12 @@
-import { Controller } from '@nestjs/common'
+import {
+  Body,
+  Controller,
+  Post,
+  Get
+} from '@nestjs/common'
 
 import { UsersService } from './users.service'
+import { User } from './entities/user.entity'
 
 @Controller('api/users')
 export class UsersController {
@@ -20,4 +26,15 @@ export class UsersController {
       return error
     }
   }
+
+  @Get('test')
+  async createUserGet(): Promise<User> {
+    try {
+      return await this.usersService.createUser({ firstName: 'a', lastName: 'b', password: 'c', email: 'e-' + Math.random() });
+    } catch (error: any) {
+      // log
+
+      return error
+    }
+  }
 }
diff --git a/src/modules/users/users.service.ts b/src/modules/users/users.service.ts
index 88cd5bf..01927e2 100644
--- a/src/modules/users/users.service.ts
+++ b/src/modules/users/users.service.ts
@@ -17,6 +17,8 @@ export class UsersService {
   ): Promise<User> {
     let user: User | null = null
 
+    // @ts-ignore
+    console.log(this.userRepository.em.getContext().id)
     // try {
       user = await this.userRepository.create(createUserDto)
       await this.userRepository.persistAndFlush(user)
```

**Reply by B4nan** [reply]
1. dont write me private messages on slack, i am not your personal support guy. i invited you there because there are others who can help. dont expect help in the exact moment you ask for it, that's not how open source work. remember you are asking for free support. 
2. you are debugging wrong places, it is correct that both `currentRequestContext` and `getEntityManager` will give you undefined, if they are not executed inside the context handler
3. same with EM id - it will be 1 if you try to work with global EM - that is the whole point of the validation error you see - you are basically checking whether the error is issued in correct circumstances. dont debug the error, debug the cause. you need to find out why the request context middleware is not working - most probably it is not even used for the route. i have no idea how this works in fastify, never used it myself.

**Comment by B4nan** [comment]
Looks like there is a bug in the nest/fastify platform or somewhere down that path, that is causing the context to be lost for requests with payload (so post/put/patch).

https://github.com/iamolegga/nestjs-pino/issues/546

(that issue is about different libraries, but the problem is the same, they also use `AsyncLocalStorage` and the context gets lost for post requests)

I was able to make it work with fastify by using nestjs request scopes instead:

```ts
import { MikroOrmModule } from '@mikro-orm/nestjs'
import { Module, Scope } from '@nestjs/common';

@Module({
  imports: [
    MikroOrmModule.forRoot({
      registerRequestContext: false, // disable the automatic context
      scope: Scope.REQUEST, // use nestjs request scopes instead
      dbName: 'test.db',
      debug: true,
      type: 'sqlite',
      entities: [__dirname + '/modules/**/entities/*.entity.js'],
      entitiesTs: [__dirname + '/modules/**/entities/*.entity.ts'],
    }),
    UsersModule
  ]
})
export class AppModule {
}
```

Note that I also used `forRoot` instead of `forRootAsync` - looks like there is a bug in `forRootAsync` for this case.

edit: with `forRootAsync` it works too, but the `scope` needs to be specified next to `useFactory`, not inside:

```ts
@Module({
  imports: [
    MikroOrmModule.forRootAsync({
      useFactory: () => ({
        registerRequestContext: false,
        dbName: 'test.db',
        debug: true,
        type: 'sqlite',
        entities: [__dirname + '/modules/**/entities/*.entity.js'],
        entitiesTs: [__dirname + '/modules/**/entities/*.entity.ts'],
      }),
      scope: Scope.REQUEST,
    }),
    UsersModule
  ]
})
export class AppModule {
}
```

**Reply by romainbrancourt** [reply]
@B4nan Hi, im having the same bug since i moved to @mikro-orm/core version 5.
It appears to happens when combining nestjs + fastify + mikro-orm v5.
Here is a minimal repo to reproduce: https://github.com/romainbrancourt/mikroorm-nest-v5.
After installing and initializing sqlite db, send a POST request to localhost:3000/
If you fallback the mikro-orm packages to version 4, it will fix the problem.
Thx in advance

**Reply by B4nan** [reply]
Well, you are replying to a post where I gave the solution, so just follow it. Again, this is a bug somewhere down the path in fastify/nest and their dependencies, nothing in MikroORM. But with nest request scopes it works.

**Reply by romainbrancourt** [reply]
Yes thanks you for the solution, i was only pointing the fact that the bug started when using v5.
It might have been related in some way.
But it seems not, thx for your work 🙂

**Reply by B4nan** [reply]
No, the bug did not start with v5. v5 added a validation, so you can see there is a bug, before you just didn't know. Nothing changed in v5 in this regard, just added validation, the problem is not in MikroORM. 

edit: to be correct, the underlying mechanism changed, we now use async local storage instead of domain api, which obviously does have effect on this. Its usage here is correct, otherwise it woulnd't work with other projects, e.g. with express it works fine, with other FWs it also works fine, using the helper itself works fine, using the context decorator also works fine. The problem is that the async context can be lost due to some other packages down the path.

**Reply by romainbrancourt** [reply]
Alright, thx for the explanation 👍

**Reply by n4an** [reply]
That is apply to current situation?

> Performance# Using request-scoped providers will have an impact on application performance. While Nest tries to cache as much metadata as possible, it will still have to create an instance of your class on each request. Hence, it will slow down your average response time and overall benchmarking result. Unless a provider must be request-scoped, it is strongly recommended that you use the default singleton scope.

**Reply by n4an** [reply]
@B4nan now is other problem if I use loggerFactory: (options: LoggerOptions) => new DatabaseLogger(options) like here https://wanago.io/2021/10/04/api-nestjs-logging-typeorm/

    this.logsService.createLog({
                     ^
TypeError: Cannot read property 'createLog' of undefined

**Reply by n4an** [reply]
@B4nan good news I hope https://github.com/nestjs/nest/issues/8837

**Comment by n4an** [comment, marked as answer]
Fixt https://github.com/fastify/fastify/blob/e9e889a21cac35e799603a006ccfed65636b390f/lib/contentTypeParser.js (edited)

**Comment by 47vigen** [comment]
I found the same issue today with the update to version 5.0.0-rc.2
I just init mikro orm and get this 
`
(node:23384) UnhandledPromiseRejectionWarning: ValidationError: Using global EntityManager instance methods for context specific actions is disallowed. If you need to 
work with the global instance's identity map, use `allowGlobalContext` configuration option or `fork()` instead.
`

init options :
`
{
  debug: !PROD,
  type: 'mongo',
  clientUrl: DATABASE_URL,
  registerRequestContext: false,
  entities: [...],
}
`

**Reply by B4nan** [reply]
Do you use nest with fastify? Did you upgrade all your dependencies?

And how do you handle request contexts, given you explicitly disable the automatic one via `registerRequestContext: false`?

> I just init mikro orm and get this

Worth noting this will never be thrown just from calling `MikroORM.init()`.

**Reply by 47vigen** [reply]
I just use fastify but I just init mikro orm 
`MikroORM.init({
  debug: !PROD,
  type: 'mongo',
  clientUrl: DATABASE_URL,
  entities: [...]
})`
and No extra!

about `registerRequestContext` I was just trying to manipulate to get results

**Reply by B4nan** [reply]
> about registerRequestContext I was just trying to manipulate to get results

Well, disabling it is exactly the way to trigger this error. How does your package.json look like?

And as said, this won't be triggered just from the init call, you need to access the global context first. Can you share full stack trace?

**Reply by 47vigen** [reply]
[https://github.com/47vigen/shopell-back-end/blob/master/src/mikro-orm.config.ts](https://github.com/47vigen/shopell-back-end/blob/master/src/mikro-orm.config.ts)

please check this config

**Reply by B4nan** [reply]
Well, start by upgrading your dependencies as I suggested. You still have the CLI package on 4.5.10, that can cause really weird issues.

https://github.com/47vigen/shopell-back-end/blob/master/package.json#L27

So where are you handing request contexts? You had `registerRequestContext` which is nestjs adapter configuration, it would have no effect in your repo. But more importantly - you need to hadnle this yourself if you dont use nest. 

You got it wrong, I see you just fork the EM once in your index.ts, that need to happen  for each request handler. I am glad I've added this validation, your project is a memory leak without it.

https://github.com/47vigen/shopell-back-end/blob/386d8e790eefc5d770e18b0fff4833c629c2af13/src/index.ts#L24

Namely this part is what you do wrong: 

https://github.com/47vigen/shopell-back-end/blob/386d8e790eefc5d770e18b0fff4833c629c2af13/src/index.ts#L41-L45

There you need to fork in the callback (and drop the fork call before, it does not make any sense (both the params are wrong, as well as signature, it is no longer 2 bool params, but an object - are you using some weak TS config? this should definitely scream at you, maybe you just ignore some TS error?)

```
    context: (request, reply) => ({
      request,
      reply,
      em: orm.em.fork(), // you dont even need the parameters for your use case, `clear` defaults to true and `useContext` is for the `RequestContext` helper you dont use
    }),
```

**Reply by 47vigen** [reply]
I did it all but again I get error
Even if I disable everything, I still have this error
`
// ** server starter function
;(async () => {
  const orm = await MikroORM.init(ormConfig)
  // const app = Fastify()

  // app.register(cors, {
  //   origin: [SERVER],
  //   credentials: true
  // })
  // app.register(cookie, {
  //   secret: COOKIE_SECRET
  // })
  // app.register(mercuriusUpload)
  // app.register(mercurius, {
  //   schema: await buildSchema({
  //     resolvers: [UserResolver, ShopResolver, CategoryResolver, ProductResolver, CartResolver],
  //     validate: true,
  //     authChecker: userAuthChecker
  //   }),
  //   context: (request, reply) => ({
  //     request,
  //     reply,
  //     em: orm.em.fork()
  //   }),
  //   graphiql: !PROD,
  //   path: '/graphql'
  // })
  // app.get('/refresh', refresh(orm.em.fork()))
  // await app.ready()

  // app.graphql.addHook('preExecution', userGraphQLHook)

  // setImmediate(() => {
  //   app.listen({ port: PORT, host: HOST }).then(() => {
  //     console.log('🚀 Server ready at http://%s:%d, in %s mode', HOST, PORT, ENV)
  //     if (!PROD) console.log('🎮 graphql playground at http://localhost:%d/graphiql !', PORT)
  //   })
  // })
})()
`

**Reply by B4nan** [reply]
I asked for the stack trace 2 times already, so once again, please show me full stack trace if you believe its the `init` call doing this.

**Reply by 47vigen** [reply]
sorry,

[discovery] ORM entity discovery started, using ReflectMetadataProvider
[discovery] - processing entity User
[discovery] - processing entity Transaction    
[discovery] - processing entity Category       
[discovery] - processing entity Shop
[discovery] - processing entity Product        
[discovery] - processing entity ProductVariable
[discovery] - processing entity ReceiveAddress
[discovery] - processing entity SendAddress   
[discovery] - processing entity Cart
[discovery] - processing entity CartItem      
[discovery] - processing entity ProductDetail 
[discovery] - processing entity Base
[discovery] - entity discovery finished, found 12 entities, took 66 ms
(node:20888) UnhandledPromiseRejectionWarning: ValidationError: Using global EntityManager instance methods for context specific actions is disallowed. If you need to 
work with the global instance's identity map, use `allowGlobalContext` configuration option or `fork()` instead.
    at Function.cannotUseGlobalContext (C:\Users\Vigen\Desktop\Projects\Shopell\back-end\node_modules\@mikro-orm\core\errors.js:83:16)
    at MongoEntityManager.getContext (C:\Users\Vigen\Desktop\Projects\Shopell\back-end\node_modules\@mikro-orm\core\EntityManager.js:650:44)
    at MongoEntityManager.getEntityFactory (C:\Users\Vigen\Desktop\Projects\Shopell\back-end\node_modules\@mikro-orm\core\EntityManager.js:638:21)
    at new UnitOfWork (C:\Users\Vigen\Desktop\Projects\Shopell\back-end\node_modules\@mikro-orm\core\unit-of-work\UnitOfWork.js:30:166)
    at new EntityManager (C:\Users\Vigen\Desktop\Projects\Shopell\back-end\node_modules\@mikro-orm\core\EntityManager.js:32:27)
    at new MongoEntityManager (C:\Users\Vigen\Desktop\Projects\Shopell\back-end\node_modules\@mikro-orm\mongodb\MongoEntityManager.js:8:1)
    at MongoDriver.createEntityManager (C:\Users\Vigen\Desktop\Projects\Shopell\back-end\node_modules\@mikro-orm\mongodb\MongoDriver.js:16:16)
    at Function.init (C:\Users\Vigen\Desktop\Projects\Shopell\back-end\node_modules\mikro-orm\node_modules\@mikro-orm\core\MikroORM.js:44:29)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:20888) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:20888) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Reply by B4nan** [reply]
You must have something wrongly setup, or some version misalignment. During discovery, the flag `allowGlobalContext ` is actually being set to false for short time, to allow the `init` method to setup things.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/MikroORM.ts#L39-L42

Your stack trace says its coming from `MikroORM.js:44:29`, but this line is an if statement in RC 2:

https://unpkg.com/browse/@mikro-orm/core@5.0.0-rc.2/MikroORM.js

This is a dependency mismatch for sure. Wipe your node modules and lock file, and reinstall. And I hope you already fixed your CLI version I pointed out in the first place.

**Reply by 47vigen** [reply]
I fixed CLI version and reinstall dependencies but I still have error
when I use v4.5 there was not any error
This error only happens when I use version 5

**Reply by B4nan** [reply]
This is validation error, not seeing it in v4 does not mean its not there, v4 did not have this validation.  

And as I said, this stack trace is definitely not from RC.2 core package. Just check the code I linked, `MikroORM.init` no longer calls `createEntityManager` method directly, there is `discoverEntities` method missing in your stack trace.

**Reply by 47vigen** [reply]
I'm really sorry I took your time
I found out that I did not update the `mikro-orm` to version 5.0.0-rc.2, after the update I found out that the `mikro-orm` is empty! and I have to import `MikroORM` from the `@mikro-orm/core`  🙂

**Reply by B4nan** [reply]
ouch totally missed that, you know if it would be _minimal_ repro, it would be easier to spot :P 

anyway, I just added validation for this situation

**Reply by n4an** [reply]
> ouch totally missed that, you know if it would be _minimal_ repro, it would be easier to spot :P
> 
> anyway, I just added validation for this situation

:)

**Comment by ayoub3bidi** [comment]
https://stackoverflow.com/questions/71117269/validation-error-using-global-entity-manager-instance-methods-for-context-speci

**Reply by B4nan** [reply]
Oh jeez, people suggesting to disable validation without understanding, and being upvoted for that... Thanks for linking this, will try to add some meaningful response there.

**Comment by yangli-io** [comment]
I read the stackoverflow and came here, thanks for putting the correct answer up. I just want to point out that the design on `allowGlobalContext` requires you to have somewhat of a good understanding on concurrent systems, sql transactions and a bit of mikroorm to really know what that variable is about. 

It's too easy to just use global and solve any immediate development issues. Sometimes the issues you get from this can only be noticed in production which makes the issue rather nasty and perhaps hard to even debug. 

I would suggest totally removing this config so that people don't use it by accident and instead perhaps add an API with a warning in the name for anyone that would need it - something along the lines of `em.dangerouslyUseGlobalContext.find()`

**Reply by B4nan** [reply]
Yeah agreed, I am planning to make the warning more informative instead of just suggesting the `allowGlobalContext`, and maybe link to the stackoverflow post or somewhere into the docs. And maybe even rename it, but that would also mean harder discoverability I guess. I could also imagine a warning printed when this is enabled, with some other toggle to hide it maybe in case you really know what you are doing.

I am not sure if removing it is a good idea, but I might be biased, as it would also mean I need to modify all the tests that depend on this. And there are a lot of them... :] 

Or maybe we should keep this available only via env vars?
